src/cluster_legacy.c

```diff
@@ -1588,16 +1627,18 @@ clusterNode *clusterLookupNode(const char *name, int length) {
 }
 
 /* Get all the nodes in my shard.
- * Note that the list returned is not computed on the fly
- * via slaveof; rather, it is maintained permanently to
- * track the shard membership and its life cycle is tied
- * to this process. Therefore, the caller must not
- * release the list. */
+ * We ensure that we maintain the master nodes only after they have been assigned with a persisted
+ * shard ID. Generate the list of nodes in a shard if they are persisted, else return
+ * NULL. The caller should release the list */
 list *clusterGetNodesInMyShard(clusterNode *node) {
-    sds s = sdsnewlen(node->shard_id, CLUSTER_NAMELEN);
-    dictEntry *de = dictFind(server.cluster->shards,s);
-    sdsfree(s);
-    return (de != NULL) ? dictGetVal(de) : NULL;
+    clusterNode *master = clusterNodeGetMaster(node);
+
+    list *l = listCreate();
+    listAddNodeTail(l, master);
+    for (int i = 0; i < master->numslaves; i++) {
+        listAddNodeTail(l, master->slaves[i]);
+    }
+    return l;
 }
```

Comment:
This operation becomes O(N) and some additional memory allocation. However, I'm not bothered a lot about it since the code flow is not on the hot path.